### PR TITLE
Update TDLib & Save voice/video notes

### PR DESCRIPTION
**Pull request by viscount-sanchez**
When I run the version from upstream, it doesn't log in and instead suggests to update app (tdlib). Had to change some code to do the update.
Also added downloading voice & video notes and fix for config.humanParams.textReadSpeedWPM.

### CODE REVIEW

**Comment by ImanolBarba** [comment]
Hey, thanks for your contribution! I will review shortly.

I am not surprised about the message you got, it's been very long since i worked on this project and an update is definitely needed.

File: auth.cpp [hunk 1]
```diff
@@ -52,10 +89,10 @@ void TelegramRecorder::onAuthStateUpdate() {
         );
       },
       [this](td_api::authorizationStateWaitRegistration &) {
-        this->sendQuery(
-          td_api::make_object<td_api::registerUser>(this->config.firstName, this->config.lastName),
-          this->createAuthQueryHandler()
-        );
+//        this->sendQuery(
```
Comment (ImanolBarba) [inline auth.cpp]: Why are we commenting the user registration here? Is it not needed after the tdlib update? How are we specifying the name and last name then?

File: auth.cpp [hunk 2]
```diff
@@ -77,21 +114,22 @@ void TelegramRecorder::onAuthStateUpdate() {
...
         );
       },
-      [this](td_api::authorizationStateWaitEncryptionKey&) {
-        // Default to an empty key (at this point who cares lmao)
-        this->sendQuery(
-          td_api::make_object<td_api::checkDatabaseEncryptionKey>(""),
-          this->createAuthQueryHandler()
-        );
+      [this](td_api::authorizationStateWaitEmailAddress&) {
+          std::cout << "Enter email: " << std::flush;
```
Comment (ImanolBarba) [inline auth.cpp]: In order to disambiguate, we should say "Enter email address" here and "Enter email code" below

File: db.cpp [hunk 3]
```diff
@@ -206,7 +206,7 @@ bool TelegramRecorder::writeMessageToDB(std::shared_ptr<td_api::message>& messag
   statement += (fileOriginID == "" ? "NULL" : "'" + fileOriginID + "'") + ",";
   statement += std::to_string(message->chat_id_) + ",";
   statement += std::to_string(senderID) + ",";
-  statement += (message->reply_to_message_id_ ? ("'" + std::to_string(message->reply_in_chat_id_) + ":" + std::to_string(message->reply_to_message_id_) + "'") : "NULL") + ",";
```
Comment (ImanolBarba) [inline db.cpp:209]: Commenting this portion should invalidate the SQL statement by leaving a dangling `,` character, I don't think this we want this

File: message_reader.cpp [hunk 4]
```diff
@@ -101,7 +101,8 @@ void TelegramRecorder::markMessageAsRead(std::shared_ptr<td_api::message>& messa
   SPDLOG_DEBUG("Marking message {} from chat {} as read", message->id_, message->chat_id_);
   td_api::object_ptr<td::td_api::viewMessages> viewMessages = td_api::make_object<td_api::viewMessages>();
   viewMessages->chat_id_ = message->chat_id_;
-  viewMessages->message_thread_id_ = message->message_thread_id_;
```
Comment (ImanolBarba) [inline message_reader.cpp:104]: Why are we commenting this line?

File: telegram_recorder.cpp [hunk 5]
```diff
@@ -338,7 +349,12 @@ void TelegramRecorder::retrieveAndWriteUserFromTelegram(td_api::int53 userID) {
       TelegramUser* user = new TelegramUser;
       user->userID = u->id_;
       user->fullName = (u->last_name_ == "" ? u->first_name_ : (u->first_name_ + " " + u->last_name_));
-      user->userName = u->username_;
+      auto usernames = join(u->usernames_->active_usernames_);
+      auto disabled_usernames = join(u->usernames_->disabled_usernames_);
+      if (!disabled_usernames.empty())  {
```
Comment (ImanolBarba) [inline telegram_recorder.cpp]: I would leave the `disabled_usernames` on a separate field in the `TelegramUser` class, with the idea of adding it to a new column in the sqlite schema. Should be easier to query the info.

**Review by ImanolBarba — CHANGES_REQUESTED** [review]

**Review by ImanolBarba — APPROVED** [review]
G2G, seems curious that TDLib now considers usernames optional, but I guess that's how they do things now

**Review by ImanolBarba — APPROVED** [review]
g2g!

**Review by ImanolBarba — APPROVED** [review]
Good catch!

**Review by ImanolBarba — APPROVED** [review]

File: db.cpp (hunk 3 above)
Comment (ImanolBarba) [reply db.cpp:209]: nvm, saw the changes further down the commit stack

File: db.cpp [hunk 6]
```diff
@@ -210,7 +210,12 @@ bool TelegramRecorder::writeMessageToDB(std::shared_ptr<td_api::message>& messag
   statement += (fileOriginID == "" ? "NULL" : "'" + fileOriginID + "'") + ",";
   statement += std::to_string(message->chat_id_) + ",";
   statement += std::to_string(senderID) + ",";
-  statement += (message->reply_to_ ? ("'" + std::to_string(message->reply_to_)  + "'") : "NULL") + ",";
+  std::string reply_to = "";
```
Comment (ImanolBarba) [inline db.cpp:213]: Yeah this is actually a much better idea, this way we can still join on message id to know who this is from, as well as better reconstruct the convo.

**Review by ImanolBarba — APPROVED** [review]

**Review by ImanolBarba — APPROVED** [review]

**Review by ImanolBarba — APPROVED** [review]

File: auth.cpp (hunk 2 above)
Comment (ImanolBarba) [reply auth.cpp]: nvm, saw we're removing this later

**Comment by ImanolBarba** [comment]
Overall good, but I would request to separate active and inactive users in the `TelegramUser` class and schema

File: auth.cpp (hunk 1 above)
Comment (viscount-sanchez) [reply auth.cpp]: API has changed and old code doesn't compile.
Perhaps I don't know all the use cases, but here is how I use this app:
1. Register tg account via official app (or buy tgdata)
2. Specify names and security settings in the official app
3. Obtain app key
4. Run this app with app key

Seems like registering user is unnecessary, but maybe there are some more use cases I'm missing.

File: message_reader.cpp (hunk 4 above)
Comment (viscount-sanchez) [reply message_reader.cpp:104]: New API doesn't have such field, unless you do static cast to certain specific type. There may be several specific types and new ones in the future versions too.
Mark as read request  works even without message_thread_id_, so I just removed it.

**Comment by viscount-sanchez** [comment]
Thank you for the feedback.
It's a good idea to put disabled usernames in a separate column, I'll do it one of these days.

File: auth.cpp (hunk 1 above)
Comment (ImanolBarba) [reply auth.cpp]: Thanks, I'll check out the new process

**Comment by ImanolBarba** [comment]
I'll try to land that change later today, feel free to send another PR for that if I end up forgetting about it :)

**Pull request merged**